goldmane/pkg/aggregator/index.go

```diff
@@ -60,42 +64,81 @@ type IndexFindOpts struct {
 	filter *proto.Filter
 }
 
-func (idx *index[E]) List(opts IndexFindOpts) []*types.Flow {
+func (idx *index[E]) List(opts IndexFindOpts) ([]*types.Flow, int) {
 	logrus.WithFields(logrus.Fields{
 		"opts": opts,
 	}).Debug("Listing flows from index")
 
 	var matchedFlows []*types.Flow
+	var totalMatchedCount int
 
-	// Find the index into the DiachronicFlow array from which to start the search based on the provided page number and limit.
-	// We need to iterate through the DiachronicFlows to find the starting point, since not every DiachronicFlow is necessarily a match
-	// for the given options.
-	var i int
-	if opts.page > 0 {
-		target := opts.page * opts.limit
-		for j := 0; j < len(idx.diachronics); j++ {
-			if idx.evaluate(idx.diachronics[j], opts) != nil {
-				i++
-			}
-			if int64(i) == target {
-				break
+	target := int(opts.page * opts.pageSize)
+
+	// Iterate through the DiachronicFlows and evaluate each one until we reach the limit or the end of the list.
+	for _, diachronic := range idx.diachronics {
+		flow := idx.evaluate(diachronic, opts)
+		if flow != nil {
+			// increment the count regardless of whether we're including the key, as we need a total matching count.
+			totalMatchedCount++
+
+			// The target represents the index of the last key on the previous page, so we need to ensure that we're above
+			// that index before we start collecting flows.
+			if totalMatchedCount > target {
+				// Only append the key if
+				if opts.pageSize == 0 || int64(len(matchedFlows)) < opts.pageSize {
+					matchedFlows = append(matchedFlows, flow)
+				}
 			}
 		}
 	}
 
+	return matchedFlows, calculatePageCount(totalMatchedCount, int(opts.pageSize))
+}
+
+// Keys retrieves a unique set of keys matching the given index options.
+func (idx *index[E]) Keys(opts IndexFindOpts) ([]E, int) {
+	logrus.WithFields(logrus.Fields{
+		"opts": opts,
+	}).Debug("Listing flows from index")
+
+	var matchedKeys []E
+	var totalMatchedCount int
+
+	target := int(opts.page * opts.pageSize)
+	var previousKey *E
+
 	// Iterate through the DiachronicFlows and evaluate each one until we reach the limit or the end of the list.
-	for ; i < len(idx.diachronics); i++ {
-		flow := idx.evaluate(idx.diachronics[i], opts)
+	for _, diachronic := range idx.diachronics {
+		flow := idx.evaluate(diachronic, opts)
 		if flow != nil {
-			matchedFlows = append(matchedFlows, flow)
-		}
-
-		if opts.limit > 0 && int64(len(matchedFlows)) == opts.limit {
-			return matchedFlows
+			key := idx.keyFunc(&diachronic.Key)
+			// If the previous key does not equal the current key we know that we haven't seen this key yet, as this
+			// is sorted list and all keys with the same value are together.
```

Comment:
I think it's worth explaining explicitly why we might have duplicate keys - it's not super intuitive, given that Key is generally something meant to be unique.

(In this case, it's because we're not really talking about a key, we're talking about a field of an object used for sorting).

Might be worth doing a pass of this whole Index file and tightening up our terminology for clarity - `keyFunc` -> `getSortValue` or something like that.

Reply:
Ya good point, I might opt for renaming key to value as suggested. I'll add a comment after that.